File: src/main/haskell/CircularBuffer.hs
{-# LANGUAGE DataKinds           #-}
{-# LANGUAGE FlexibleContexts    #-}
{-# LANGUAGE OverloadedStrings   #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeOperators       #-}
{-# OPTIONS_GHC -fplugin=Language.Java.Inline.Plugin #-}
module CircularBuffer where

import           Data.Int
import           Foreign.JNI.Types          (J, JObject, JVM, jnull)
import qualified Language.Haskell.TH.Syntax as TH
import qualified Language.Java              as Java
import           Language.Java.Inline

newtype CircularBuffer a =
  CircularBuffer (J (Java.Generic (Java.Class "example.CircularBuffer") '[ Java.Interp a]))
  deriving (Eq)

new :: Int -> IO (CircularBuffer a)
new capacity = do
  obj :: J (Java.Class "example.CircularBuffer") <- Java.new
    [Java.coerce (fromIntegral capacity :: Int32)]
  return (CircularBuffer (Java.unsafeCast (Java.generic obj)))

put :: Java.Reflect a => CircularBuffer a -> a -> IO ()
put (CircularBuffer obj) x = do
  value <- Java.reflect x
  Java.call (Java.unsafeUngeneric obj) "put" [Java.coerce (Java.upcast value)]

get :: Java.Reify a => CircularBuffer a -> IO (Maybe a)
get (CircularBuffer obj) = do
  x :: JObject <- Java.call (Java.unsafeUngeneric obj) "get" []
  if x == jnull
    then return Nothing
    else Just <$> Java.reify (Java.unsafeCast x)

size :: CircularBuffer a -> IO Int
size (CircularBuffer obj) = do
  s :: Int32 <- Java.call (Java.unsafeUngeneric obj) "size" []
  return (fromIntegral s)

File: src/main/java/example/CircularBuffer.java
package example;

public class CircularBuffer<T> {
    private final T[] buf;
    private int in;
    private int out;

    public CircularBuffer(int capacity) {
        this.buf = (T[]) new Object[capacity + 1];
    }

    public synchronized void put(T x) {
        buf[in] = x;
        this.in = (this.in + 1) % this.buf.length;
    }

    public synchronized T get() {
        T x = buf[out];
        this.out = (this.out + 1) % this.buf.length;
        return x;
    }

    public synchronized int size() {
        int diff = this.in - this.out + this.buf.length;
        return diff == 0 ? 0 : diff % this.buf.length;
    }
}
